File: aop/aop-annotation-processor/src/test/java/ru/tinkoff/kora/aop/annotation/processor/AbstractTestAnnotationAspect.java
package ru.tinkoff.kora.aop.annotation.processor;

import com.palantir.javapoet.ClassName;
import com.palantir.javapoet.CodeBlock;
import com.palantir.javapoet.TypeName;
import ru.tinkoff.kora.annotation.processor.common.AnnotationUtils;

import javax.lang.model.element.ExecutableElement;
import javax.lang.model.type.TypeKind;
import java.util.List;
import java.util.Objects;
import java.util.Set;

public abstract class AbstractTestAnnotationAspect implements KoraAspect {
    protected abstract ClassName testAnnotation();

    @Override
    public Set<String> getSupportedAnnotationTypes() {
        return Set.of(testAnnotation().canonicalName());
    }

    @Override
    public ApplyResult apply(ExecutableElement executableElement, String superCall, AspectContext aspectContext) {
        var annotation = executableElement.getParameters().stream()
            .map(v -> AnnotationUtils.findAnnotation(v, testAnnotation()))
            .filter(Objects::nonNull)
            .findFirst()
            .orElse(null);
        if (annotation == null) {
            annotation = AnnotationUtils.findAnnotation(executableElement, testAnnotation());
        }
        if (annotation == null) {
            annotation = Objects.requireNonNull(AnnotationUtils.findAnnotation(executableElement.getEnclosingElement(), testAnnotation()));
        }
        var field = aspectContext.fieldFactory().constructorParam(
            ClassName.get(TestMethodCallListener.class),
            List.of()
        );
        var annotationValue = AnnotationUtils.<String>parseAnnotationValueWithoutDefault(annotation, "value");
        var b = CodeBlock.builder()
            .addStatement("this.$N.before($S)", field, annotationValue);
        if (!executableElement.getReturnType().getKind().equals(TypeKind.VOID)) {
            b.addStatement("$T _result = null", TypeName.get(executableElement.getReturnType()).box())
                .beginControlFlow("try")
                .addStatement("_result = $L($L)", superCall, executableElement.getParameters().stream().map(e -> CodeBlock.of("$N", e.getSimpleName())).collect(CodeBlock.joining(",")))
                .addStatement("this.$N.after($S, _result)", field, annotationValue);
            b.addStatement("return _result");
        } else {
                b.beginControlFlow("try")
                .addStatement("$L($L)", superCall, executableElement.getParameters().stream().map(e -> CodeBlock.of("$N", e.getSimpleName())).collect(CodeBlock.joining(",")))
                .addStatement("this.$N.after($S, null)", field, annotationValue);
        }
        b.nextControlFlow("catch (Throwable e)")
            .addStatement("this.$N.thrown($S, e)", field, annotationValue)
            .addStatement("throw e")
            .endControlFlow();

        return new ApplyResult.MethodBody(b.build());
    }
}

File: aop/aop-symbol-processor/src/test/kotlin/ru/tinkoff/kora/aop/ksp/AbstractTestAnnotationAspect.kt
package ru.tinkoff.kora.aop.ksp

import com.google.devtools.ksp.symbol.KSFunctionDeclaration
import com.squareup.kotlinpoet.ClassName
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.asClassName
import com.squareup.kotlinpoet.ksp.toTypeName
import ru.tinkoff.kora.aop.symbol.processor.KoraAspect
import ru.tinkoff.kora.aop.symbol.processor.KoraAspect.AspectContext
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findAnnotation
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findValueNoDefault

abstract class AbstractTestAnnotationAspect : KoraAspect {
    protected abstract fun testAnnotation(): ClassName

    override fun getSupportedAnnotationTypes(): Set<String> {
        return setOf(testAnnotation().canonicalName)
    }

    override fun apply(ksFunction: KSFunctionDeclaration, superCall: String, aspectContext: AspectContext): KoraAspect.ApplyResult {
        var annotation = ksFunction.parameters.asSequence()
            .mapNotNull { it.findAnnotation(testAnnotation()) }
            .firstOrNull()
        if (annotation == null) {
            annotation = ksFunction.findAnnotation(testAnnotation())
        }
        if (annotation == null) {
            annotation = ksFunction.parentDeclaration!!.findAnnotation(testAnnotation())!!
        }
        val field: String = aspectContext.fieldFactory.constructorParam(
            TestMethodCallListener::class.asClassName(),
            listOf()
        )

        val annotationValue = annotation.findValueNoDefault<String>("value")
        val b = CodeBlock.builder()
            .addStatement("this.%N.before(%S)", field, annotationValue)
        b.addStatement("var _result: %T = null", ksFunction.returnType?.toTypeName()?.copy(true))
            .beginControlFlow("try")
            .addStatement("_result = %L", superCall(superCall, ksFunction.parameters.map { it.name!!.asString() }))
            .addStatement("this.%N.after(%S, _result)", field, annotationValue)
        b.addStatement("return _result")
        b.nextControlFlow("catch (e: Throwable)")
            .addStatement("this.%N.thrown(%S, e)", field, annotationValue)
            .addStatement("throw e")
            .endControlFlow()

        return KoraAspect.ApplyResult.MethodBody(b.build())
    }
}
